packages/cw-orch-fns-derive/src/query_fns.rs

```diff
@@ -157,5 +157,137 @@ pub fn query_fns_derive(input: ItemEnum) -> TokenStream {
+                quote!(
+                    #variant_doc
+                    async fn #variant_func_name(&self) -> ::core::result::Result<#response, ::cw_orch::prelude::CwOrchError> {
+                        let msg = #name::#variant_name;
+                        self.async_query(&msg #maybe_into).await
+                    }
+                )
+            },
+            Fields::Named(variant_fields) => {
+                if is_attributes_sorted{
+                    // sort fields on field name
+                    LexiographicMatching::default().visit_fields_named_mut(variant_fields);
+                }
+
+                // remove attributes from fields
+                variant_fields.named.iter_mut().for_each(|f| f.attrs = vec![]);
+
+                // Parse these fields as arguments to function
+                let variant_fields = variant_fields.named.clone();
+                let variant_idents = variant_fields.iter().map(|f|f.ident.clone().unwrap());
+
+                let variant_attr = variant_fields.iter();
+                quote!(
+                    #variant_doc
+                    #[allow(clippy::too_many_arguments)]
+                    async fn #variant_func_name(&self, #(#variant_attr,)*) -> ::core::result::Result<#response, ::cw_orch::prelude::CwOrchError> {
+                        let msg = #name::#variant_name {
+                            #(#variant_idents,)*
+                        };
+                        self.async_query(&msg #maybe_into).await
+                    }
+                )
+            }
+        }
+    });
+
+    let derived_trait = quote!(
+        /// Automatically derived trait that allows you to call the variants of the message directly without the need to construct the struct yourself.
+        pub trait #bname<Chain: ::cw_orch::environment::AsyncWasmQuerier + ::cw_orch::environment::ChainState, #type_generics>: ::cw_orch::prelude::AsyncCwOrchQuery<Chain, QueryMsg = #entrypoint_msg_type #ty_generics > #where_clause {
+            #(#variant_fns)*
+        }
+    );
+
+    // We need to merge the where clauses (rust doesn't support 2 wheres)
+    // If there is no where clause, we simply add the necessary where
+    let necessary_where = quote!(SupportedContract: ::cw_orch::prelude::AsyncCwOrchQuery<Chain, QueryMsg = #entrypoint_msg_type #ty_generics >);
+    let combined_where_clause = where_clause
+        .map(|w| {
+            quote!(
+                #w #necessary_where
+            )
+        })
+        .unwrap_or(quote!(
+            where
+                #necessary_where
+        ));
+
+    let derived_trait_impl = quote!(
+        #[automatically_derived]
+        impl<SupportedContract, Chain: ::cw_orch::environment::AsyncWasmQuerier + ::cw_orch::environment::ChainState, #type_generics> #bname<Chain, #type_generics> for SupportedContract
+        #combined_where_clause {}
+    );
+
+    let expand = quote!(
+        #derived_trait
+
+        #derived_trait_impl
```

Comment:
For instance here you need to target-flag this derive_trait_impl just like for the sync case. Better merge those into a common function